airbyte-connector-builder-server/connector_builder/impl/default_api.py

```diff
@@ -144,11 +148,34 @@ async def read_stream(self, stream_read_request_body: StreamReadRequestBody = Bo
                 detail=f"Could not perform read with with error: {error.args[0]}",
             )
 
-        return StreamRead(logs=log_messages, slices=[single_slice],
-                          inferred_schema=schema_inferrer.get_stream_schema(stream_read_request_body.stream))
+        return StreamRead(
+            logs=log_messages, slices=[single_slice], inferred_schema=schema_inferrer.get_stream_schema(stream_read_request_body.stream)
+        )
+
+    async def resolve_manifest(
+        self, manifest_resolve_request_body: ManifestResolveRequestBody = Body(None, description="")
+    ) -> ManifestResolve:
+        """
+        Using the provided manifest, resolves $refs and $options and returns the resulting manifest to the client.
+        :param manifest_resolve_request_body: Input manifest whose $refs and $options will be resolved
+        :return: Airbyte record messages produced by the sync grouped by slice and page
+        """
+        try:
+            return ManifestResolve(
+                manifest=ManifestDeclarativeSource(
+                    manifest_resolve_request_body.manifest, construct_using_pydantic_models=True
+                ).resolved_manifest()
+            )
+        except Exception as error:
+            self.logger.error(f"Could not resolve manifest with with error: {error.args[0]} - {self._get_stacktrace_as_string(error)}")
+            raise HTTPException(
+                status_code=400,
```

Comment:
It seems weird to me to assume that this is a user input error (i.e. a `400` error) and not an internal server error. I know @lmossman wanted to track stream test errors. Will this enable us to do this? 

Note: This can probably be scoped out of the PR. My main goal is to bring awareness and make sure we are aligned on short term incoming changes

Reply:
Does the exception thrown by the `ManifestDeclarativeSource.resolved_manifest()` method provide any indication as to what type of error occurred? Ideally this would return a 400 if the manifest is invalid in some way, but it would return a 500 if some other unrelated issue happened (e.g. some bug in the code causing an uncaught exception)

Reply:
Currently the `4**` errors are returned for invalid manifests & requests, and an unhandled exception in the code would return a 500, openapi's default.

I took a closer look at the errors surfaced by `resolved_manifest`, which are all raised during initialization of `ManifestDeclarativeSource`. They have varying levels of usefulness - a manifest with an unresolvable reference will return an error stating such (as in the `test_resolve_manifest_unresolvable_references` test case), but manifests that are invalid for other reasons (e.g. missing some fields, as in the `test_resolve_manifest_invalid` test case) may have more generic error messages. If there is a category of errors that you'd like special handling for I'm happy to add it.